pkg/blobstore/local/file_backed_location_record_array.go

```diff
@@ -0,0 +1,88 @@
+package local
+
+import (
+	"encoding/binary"
+	"sync"
+
+	"github.com/buildbarn/bb-storage/pkg/filesystem"
+)
+
+func serializeLocationRecord(locationRecord LocationRecord) []byte {
+	serialised := make([]byte, 60)
+	for i := 0; i < 32; i++ {
+		serialised[i] = locationRecord.Key.Digest[i]
+	}
+	binary.LittleEndian.PutUint32(serialised[32:36], locationRecord.Key.Attempt)
+	binary.LittleEndian.PutUint64(serialised[36:44], uint64(locationRecord.Location.BlockID))
+	binary.LittleEndian.PutUint64(serialised[44:52], uint64(locationRecord.Location.OffsetBytes))
+	binary.LittleEndian.PutUint64(serialised[52:60], uint64(locationRecord.Location.SizeBytes))
+
+	return serialised
+}
+
+func deserializeLocationRecord(record []byte) LocationRecord {
+	var keyDigest [32]byte
+	copy(keyDigest[0:32], record[0:32])
+	keyAttempt := binary.LittleEndian.Uint32(record[32:36])
+	locationBlockID := int(binary.LittleEndian.Uint64(record[36:44]))
+	locationOffset := int64(binary.LittleEndian.Uint64(record[44:52]))
+	locationSize := int64(binary.LittleEndian.Uint64(record[52:60]))
+
+	key := LocationRecordKey{
+		Digest:  keyDigest,
+		Attempt: keyAttempt,
+	}
+
+	location := Location{
+		BlockID:     locationBlockID,
+		OffsetBytes: locationOffset,
+		SizeBytes:   locationSize,
+	}
+
+	return LocationRecord{
+		Key:      key,
+		Location: location,
+	}
+}
+
+type fileBackedLocationRecordArray struct {
+	recordFile filesystem.FileReadWriter
+	lock       sync.Mutex
+}
+
+// NewFileBackedLocationRecordArray creates a persistent LocationRecordArray.
+// Works by using a file as an array-like structure, writing serialised
+// LocationRecords next to each other and calculating where to Read/Write using
+// the length of a serialised record multiplied by the index.
+func NewFileBackedLocationRecordArray(recordFile filesystem.FileReadWriter) LocationRecordArray {
```

Comment:
What's a bit annoying is that `blockdevice.BlockDevice` does not implement a `filesystem.FileReadWriter`. This is a shame, because FileBackedLocationRecordArray would work perfectly fine with a block device, while the current implementation currently doesn't.

What are your thoughts on renaming this to BlockDeviceBackedLocationRecordArray and letting it use `blockdevice.BlockDevice`? It's all right to refer to it as being 'file backed' in the configuration file. It's just that it could also be block device backed if we wanted to.

Reply:
I don't particularly mind mandating the use of a block device for the backend, but doing this is a little inflexible. Is there any reason not to implement a `filesystem.FileReadWriter` for a block device (I guess it's a bit of an oxymoron)?

Alternatively, could we not use a less strict interface (I think a ReadWriterAt would be sufficient) and allow either?

Reply:
You can't Truncate() a block device. Yes, sorry. The interface is already called blockdevice.ReadWriterAt; not blockdevice.BlockDevice.

Reply:
I've updated this to use a `blockdevice.ReadWriterAt`, with it being configurable whether to use a file or a block device for the storage. If preferred, I can make this block-device only and automatically calculate the size of the map (or do this in the block device case). I also wasn't sure if it might be better to just detect whether the path points to a block device or not, which is another option.